service/worker/indexer/esProcessor.go

```diff
@@ -0,0 +1,265 @@
+	}
+
+	// ElasticBulkProcessor is interface for elastic.BulkProcessor
+	// (elastic package doesn't provide such interface that tests can mock)
+	ElasticBulkProcessor interface {
+		Start(ctx context.Context) error
+		Stop() error
+		Close() error
+		Stats() elastic.BulkProcessorStats
+		Add(request elastic.BulkableRequest)
+		Flush() error
+	}
+
+	// esProcessorImpl implements ESProcessor, it's an agent of elastic.BulkProcessor
+	esProcessorImpl struct {
+		processor     ElasticBulkProcessor
+		mapToKafkaMsg collection.ConcurrentTxMap // used to map ES request to kafka message. key: esDocID+visType
+		config        *Config
+		logger        bark.Logger
+		metricsClient metrics.Client
+	}
+)
+
+var _ ESProcessor = (*esProcessorImpl)(nil)
+
+const (
+	// retry configs for es bulk processor
+	esProcessorInitialRetryInterval = 200 * time.Millisecond
+	esProcessorMaxRetryInterval     = 20 * time.Second
+)
+
+// NewESProcessorAndStart create new ESProcessor and start
+func NewESProcessorAndStart(config *Config, client *elastic.Client, processorName string,
+	logger bark.Logger, metricsClient metrics.Client) (ESProcessor, error) {
+	p := &esProcessorImpl{
+		config: config,
+		logger: logger.WithFields(bark.Fields{
+			logging.TagWorkflowComponent: logging.TagValueIndexerESProcessorComponent,
+		}),
+		metricsClient: metricsClient,
+	}
+
+	processor, err := client.BulkProcessor().
+		Name(processorName).
+		Workers(config.ESProcessorNumOfWorkers()).
+		BulkActions(config.ESProcessorBulkActions()).
+		BulkSize(config.ESProcessorBulkSize()).
+		FlushInterval(config.ESProcessorFlushInterval()).
+		Backoff(elastic.NewExponentialBackoff(esProcessorInitialRetryInterval, esProcessorMaxRetryInterval)).
+		After(p.bulkAfterAction).
+		Do(context.Background())
+	if err != nil {
+		return nil, err
+	}
+
+	p.processor = processor
+	p.mapToKafkaMsg = collection.NewShardedConcurrentTxMap(1024, p.hashFn)
+	return p, nil
+}
+
+func (p *esProcessorImpl) Stop() {
+	p.processor.Stop()
+	p.mapToKafkaMsg = nil
+}
+
+// Add an ES request, and an map item for kafka message
+func (p *esProcessorImpl) Add(request elastic.BulkableRequest, key string, kafkaMsg messaging.Message) {
+	if p.mapToKafkaMsg.Contains(key) {
+		kafkaMsg.Ack() // duplicate message
+		return
+	}
+	p.mapToKafkaMsg.Put(key, kafkaMsg)
+	p.processor.Add(request)
+}
+
+// bulkAfterAction is triggered after bulk processor commit
+func (p *esProcessorImpl) bulkAfterAction(id int64, requests []elastic.BulkableRequest, response *elastic.BulkResponse, err error) {
+	if err != nil {
+		// This happens after configured retry, which means something bad happens on cluster or index
+		// When cluster back to live, processor will re-commit those failure requests
+		p.logger.WithFields(bark.Fields{
+			logging.TagErr: err,
+		}).Error("Error commit bulk request.")
+
+		for _, request := range requests {
+			p.logger.WithFields(bark.Fields{
+				logging.TagESRequest: request.String(),
```

Comment:
Will this be too noisy?

Reply:
It shouldn't be that noisy. 
This err section happens when the bulk request returned with err, which means all requests in this bulk failed. This may be caused by ES cluster issue, in which case we log all requests in bulk (only one time) so that later when ES is back online, we can check if these error requests been resent successfully.